staging/src/k8s.io/client-go/tools/cache/reflector_watchlist_test.go

```diff
@@ -0,0 +1,521 @@
+			listWatcher.closeAfterListRequests = scenario.closeAfterListRequests
+			if scenario.disableUseWatchList {
+				reflector.UseWatchList = false
+			}
+
+			err := reflector.ListAndWatch(stopCh)
+			if scenario.expectedError != nil && err == nil {
+				t.Fatalf("expected error %q, got nil", scenario.expectedError)
+			}
+			if scenario.expectedError == nil && err != nil {
+				t.Fatalf("unexpected error: %v", err)
+			}
+			if scenario.expectedError != nil && err.Error() != scenario.expectedError.Error() {
+				t.Fatalf("expected error %q, got %q", scenario.expectedError, err.Error())
+			}
+
+			verifyWatchCounter(t, listWatcher, scenario.expectedWatchRequests)
+			verifyListCounter(t, listWatcher, scenario.expectedListRequests)
+			verifyRequestOptions(t, listWatcher, scenario.expectedRequestOptions)
+			verifyStore(t, store, scenario.expectedStoreContent)
+		})
+	}
+}
+
+func verifyRequestOptions(t *testing.T, lw *fakeListWatcher, expectedRequestOptions []metav1.ListOptions) {
+	if len(lw.requestOptions) != len(expectedRequestOptions) {
+		t.Fatalf("expected to receive exactly %v requests, got %v", len(expectedRequestOptions), len(lw.requestOptions))
+	}
+
+	for index, expectedRequestOption := range expectedRequestOptions {
+		actualRequestOption := lw.requestOptions[index]
+		if actualRequestOption.TimeoutSeconds == nil && expectedRequestOption.TimeoutSeconds != nil {
+			t.Fatalf("expected the request to specify TimeoutSeconds option but it didn't, actual = %#v, expected = %#v", actualRequestOption, expectedRequestOption)
+		}
+		if actualRequestOption.TimeoutSeconds != nil && expectedRequestOption.TimeoutSeconds == nil {
+			t.Fatalf("unexpected TimeoutSeconds option specified, actual = %#v, expected = %#v", actualRequestOption, expectedRequestOption)
+		}
+		// ignore actual values
+		actualRequestOption.TimeoutSeconds = nil
+		expectedRequestOption.TimeoutSeconds = nil
+		if !equality.Semantic.DeepEqual(actualRequestOption, expectedRequestOption) {
+			t.Fatalf("expected %#v, got %#v", expectedRequestOption, actualRequestOption)
+		}
+	}
+}
+
+func verifyListCounter(t *testing.T, lw *fakeListWatcher, expectedListCounter int) {
+	if lw.listCounter != expectedListCounter {
+		t.Fatalf("unexpected number of LIST requests, got: %v, expected: %v", lw.listCounter, expectedListCounter)
+	}
+}
+
+func verifyWatchCounter(t *testing.T, lw *fakeListWatcher, expectedWatchCounter int) {
+	if lw.watchCounter != expectedWatchCounter {
+		t.Fatalf("unexpected number of WATCH requests, got: %v, expected: %v", lw.watchCounter, expectedWatchCounter)
+	}
+}
+
+func verifyStore(t *testing.T, s Store, expectedPods []v1.Pod) {
+	rawPods := s.List()
+	if len(rawPods) != len(expectedPods) {
+		t.Fatalf("expect to find %v items in the storage but found %v", len(expectedPods), len(rawPods))
+	}
+
+	actualPods := []v1.Pod{}
+	for _, p := range rawPods {
+		actualPods = append(actualPods, *p.(*v1.Pod))
+	}
+
+	for _, expectedPod := range expectedPods {
```

Comment:
instead of this, let's just sort the both lists by name and compare the whole lists - it's more intuitive.

[This code in general would return true for the following lists:
- [a, a, b]
- [a, b, b]